bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -4362,6 +4478,628 @@ void ReplicaImp::setConflictDetectionBlockId(const ClientRequestMsg &clientReqMs
+
+  if (numOfRequests > 0) {
+    // TODO(GG): should be verified in the validation of the PrePrepare . Consdier to remove this assert
+    ConcordAssert(!config_.timeServiceEnabled || time_service_manager_->hasTimeRequest(*ppMsg));
+
+    histograms_.numRequestsInPrePrepareMsg->record(numOfRequests);
+    Bitmap requestSet(numOfRequests);
+    size_t reqIdx = 0;
+    uint16_t numOfSpecialReqs = 0;
+    RequestsIterator reqIter(ppMsg);
+    char *requestBody = nullptr;
+
+    bool shouldRunRequestsInParallel =
+        false;  // true IFF we have requests that will be executed in parallel to the main replica thread
+
+    //////////////////////////////////////////////////////////////////////
+    // Phase 1:
+    // a. Find the requests that should be executed
+    // b. Send reply for each request that has already been executed
+    // c. skip over invalid requests
+    // d. count number of "special requests"
+    //////////////////////////////////////////////////////////////////////
+
+    if (!recoverFromErrorInRequestsExecution) {
+      markSpecialRequests(reqIter,
+                          requestBody,
+                          numOfSpecialReqs,
+                          reqIdx,
+                          requestSet,
+                          allowParallelExecution,
+                          shouldRunRequestsInParallel);
+    } else {  // if we recover
+      requestSet = mapOfRequestsThatAreBeingRecovered;
+      // count numOfSpecialReqs
+      while (reqIter.getAndGoToNext(requestBody)) {
+        ClientRequestMsg req((ClientRequestMsgHeader *)requestBody);
+        if (config_.timeServiceEnabled && (req.flags() & MsgFlag::TIME_SERVICE_FLAG) && reqIdx == 0) {
+          numOfSpecialReqs++;
+          reqIdx++;
+          continue;
+        }
+        // TODO(GG): should be verified in the validation of the PrePrepare . Consdier to remove this assert
+        ConcordAssert(!(req.flags() & MsgFlag::TIME_SERVICE_FLAG));
+        if ((req.flags() & MsgFlag::KEY_EXCHANGE_FLAG) || (req.flags() & MsgFlag::RECONFIG_FLAG)) {
+          numOfSpecialReqs++;
+          reqIdx++;
+          continue;
+        }
+        reqIdx++;
+      }
+      reqIter.restart();
+    }
+    executeAllPreprepareRequests(allowParallelExecution,
+                                 shouldRunRequestsInParallel,
+                                 numOfSpecialReqs,
+                                 ppMsg,
+                                 requestSet,
+                                 recoverFromErrorInRequestsExecution);
+
+  } else  // if we don't have requests in ppMsg
+  {
+    // send internal message that will call to finishExecutePrePrepareMsg
+    ConcordAssert(activeExecutions_ == 0);
+    activeExecutions_ = 1;
+    InternalMessage im = FinishExecutePrePrepareInternalMsg{ppMsg, nullptr};  // TODO(GG): check....
+    getIncomingMsgsStorage().pushInternalMsg(std::move(im));
+  }
+}
+
+void ReplicaImp::markSpecialRequests(RequestsIterator &reqIter,
```

Comment:
Should also handle CLIENTS_PUB_KEYS_FLAG type of requests

Reply:
Done.